ledger/db/src/lib.rs

```diff
@@ -539,36 +539,51 @@ impl LedgerDB {
         block: &Block,
         block_contents: &BlockContents,
     ) -> Result<(), Error> {
-        // Check that version is correct
-        if block.version != BLOCK_VERSION {
-            return Err(Error::InvalidBlock);
-        }
-
-        // A block must have outputs.
-        if block_contents.outputs.is_empty() {
-            return Err(Error::NoOutputs);
-        }
-
-        // Non-origin blocks must have key images.
-        if block.index != 0 && block_contents.key_images.is_empty() {
-            return Err(Error::InvalidBlock);
-        }
-
+        // Check version is correct
         // Check if block is being appended at the correct place.
         let num_blocks = self.num_blocks()?;
         if num_blocks == 0 {
             // This must be an origin block.
+
+            // The origin block is version 0
+            if block.version != 0 {
+                return Err(Error::InvalidBlock);
+            }
+
+            // The origin block is index '0' with default-initialized parent ID, by
+            // convention
             if block.index != 0 || block.parent_id != BlockID::default() {
                 return Err(Error::InvalidBlock);
             }
         } else {
-            // The block must have the correct index and parent.
             let last_block = self.get_block(num_blocks - 1)?;
+
+            // The block's version should be bounded by
+            // [prev block version, min(prev block version + 1, max block version)]
+            if block.version >= last_block.version && block.version <= BLOCK_VERSION {
+                if block.version > last_block.version + 1 {
+                    return Err(Error::InvalidBlock);
+                }
+            } else {
+                return Err(Error::InvalidBlock);
+            }
```

Comment:
Should we allow version skipping? Instead bound by
```
// [prev block version, max block version]
```

If hardfork versions are consensed on, then without version skipping the network would have a hard time navigating buggy releases. If there is a scheduled hardfork but the code has a bug, then it would be hard to replace the new code with different code that has the same version number. When nodes consense on the new version number, if half of them have the old buggy code and half have the fixed code, then the network will get stuck on the next block because nodes will have different enclaves and no quorum will exist on the same enclave.

Right now that is not a problem because upgrading the network is a 'point in time' event where node software can be actively coordinated.

Reply:
I think that right now, and for as long as upgrading happens in an all-nodes-or-none fashion, this is fine as is. If we push out a version that is unable to produce blocks then we will roll back to the old version and try again.

Reply:
If we might need to remove this in the future, is it better to never have it in the first place?

Reply:
I have no preference.

Reply:
Ok I will just remove it

Reply:
👍 
Note that it will likely break the test I added... I'll fix it.

Reply:
Actually the test can't test for that yet because of the hardcoded bound on `BLOCK_VERSION` which is currently `1`. Oh well.

Reply:
Thank you!

Reply:
In the event of a problem with the network where no new transactions have made it through, we would roll back.